jolt-core/src/jolt/vm/ram.rs

```diff
@@ -359,30 +371,60 @@ impl<F: JoltField, ProofTranscript: Transcript> BatchableSumcheckInstance<F, Pro
                     *x -= *y;
                 });
 
-            // If transitioning to phase 2, prepare H
+            // If transitioning to phase 2, prepare H polynomials
             if round == K_log - 1 {
                 prover_state.eq_r_r = prover_state.B.final_sumcheck_claim();
 
-                // Compute H using the final F values
+                // Compute H polynomials for each decomposed part
                 let trace = self.trace.as_ref().expect("Trace not set");
                 let memory_layout = self.memory_layout.as_ref().expect("Memory layout not set");
+                let d = prover_state.d;
+                let chunk_sizes = &prover_state.chunk_sizes;
+
+                let mut H_polys = Vec::with_capacity(d);
+
+                for i in 0..d {
+                    let H_vec: Vec<F> = trace
+                        .iter()
+                        .map(|cycle| {
+                            let address =
+                                remap_address(cycle.ram_access().address() as u64, memory_layout)
+                                    as usize;
+
+                            // Decompose address to get the i-th chunk
+                            let mut remaining_address = address;
+                            for j in 0..d {
+                                let chunk_modulo = 1 << chunk_sizes[d - 1 - j];
+                                let chunk_value = remaining_address % chunk_modulo;
+                                remaining_address /= chunk_modulo;
+
+                                if j == i {
+                                    return prover_state.F[chunk_value];
+                                }
+                            }
+                            F::zero() // Should never reach here
```

Comment:
idk if this is exactly right but I think this is a little cleaner
```suggestion
                            // Decompose address to get the i-th chunk
                            let (left, right) = chunk_sizes.split_at(d - j);
                            let shift = right.iter().sum();
                            let chunk_size = left.last().unwrap();
                            let address_chunk = (address >> shift) % chunk_size;
                            prover_state.F[address_chunk]
```